feat(development): add PR review guidelines (#107)
ChainSafe/engineering-handbook: docs/3_development/1_development-flow/3_review.md

- Ensure readability is retained in code e.g. complex nested ternary operators should have a comment note

> - Ensure adequate testing exists where needed

Reviewer: I have a need to define `adequate` 🙈 but realized that is out of scope - maybe just link to QA docs?

Author: https://github.com/ChainSafe/engineering-handbook/pull/107/commits/4781b6d47a7e258410bf1a82d7dabe9bdf9e09e9

> - Code change requests include a summary of work which outlines the objectives, decisions & considerations for the code change. Examples:

Reviewer: Do we want to link certain PR templates or at least mention as good to have?

    - This PR's objective was to implement middleware for logging, this was completed however presentation & storage of logs is incomplete pending discussion. Certain components are outside the logging scope at the moment.
    - This PR includes minor styling tweaks on the desktop & mobile navigation menus, refactored navigation list logic.

## Communication

> Ask a teammate on the relevant team channel for a review, this ensures the request is clear as well as gives transparency to the rest of your team if commentary or considerations need to be made.

Reviewer: ```suggestion
While platforms like Github allow selecting peer reviewers, additionally ask a teammate on the relevant team communication channel for a review. This ensures the request is clear and gives transparency to the rest of your team if commentary or considerations need to be made.
```

Reviewer: Actually, makes sense that `Communication` part is under `Requesting reviews`

Author: https://github.com/ChainSafe/engineering-handbook/pull/107/commits/2f85bd53bf75c909c33e7c68adb32a20f2d6f5c1

> ## Walkthroughs

Reviewer: ```Walkthroughs are peer programming for reviews. If possible or required, ask for a walkthrough within your team. If your teammate has the capacity, reviewing together allows you to share the context of your work with your reviewer.```

> If a project isn't setup with [codeowners](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners) when you open a pull request, and mark it ready for review, you'll need to get your peers to review it. Every repository should have two different GitHub teams added to it (at minimum): `project_name-admins` and `project_name`, request the **non admin** team for review will allow your whole team to receive a notification about your new pull request.

Reviewer: Suggesting structure changes:
```
If a project isn't setup with [codeowners](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners) when you open a pull request, and mark it ready for review, you'll need to get your peers to review it. At a minimum, every repository should have two different GitHub teams added to it:
- `project_name-admins` - // define what it means
- `project_name` - // define what it means

Requesting the **non-admin** team for review will allow your whole team to receive a notification about your new pull request.
```

Author: https://github.com/ChainSafe/engineering-handbook/pull/107/commits/2f85bd53bf75c909c33e7c68adb32a20f2d6f5c1

## Reviewer guide

> The guide for reviewers is mostly a condensed version of [Google’s code review best practices](https://google.github.io/eng-practices/review/reviewer/). Some parts are just c/p, some parts are emphasized, some things are left out, and some are added or are different.

Reviewer: ```suggestion
The following guide for reviewers below is mostly a condensed and customized version of [Google’s code review best practices](https://google.github.io/eng-practices/review/reviewer/). For additional context we recommend reading the full guide but here are the most important points.
```

Author: https://github.com/ChainSafe/engineering-handbook/pull/107/commits/514d590f3a27b4d7d1e74de5dcf488619e1b27a2

### Speed of the code review

#### Why should code reviews be fast

Because the team is the most important! If code reviews are slow, few things will happen for sure:

- the author will start working on something else while waiting, and keep forgetting the context
- the author will finally get the reviewer’s feedback, but now he is in the middle of something else. Days will creep in.
- then the author will need to refresh his memory and address changes, making the reviewer forget about the context.
- the team will lose time and small cracks of frustration and unhappiness will appear in the team.

This is only worse if more reviewers are assigned. We will waste significant time on this context switching. It is a very inefficient process!

#### How fast should code reviews be

If we are not in the middle of a focused task, we should do a code review shortly after it comes in.

One business day is the maximum time it should take to respond to a code review request (i.e., the first thing the next morning).

Following these guidelines means that a typical code change request should get multiple rounds of review (if needed) within a single day.

> 📢 If we are in the middle of a focused task, such as writing code, we shouldn’t interrupt ourselves to do a code review. Instead, we should wait for a breakpoint in our work to respond to code change request.

Reviewer: ```suggestion
:::note
If we are in the middle of a focused task, such as writing code, we shouldn’t interrupt ourselves to do a code review. Instead, we should wait for a breakpoint in our work to respond to code change request.
:::
```

#### Fast responses

When we talk about fast code reviews, we are mainly talking about the response time. We should take time when reviewing the actual code change request. We need to fully understand the code. Ideally, the whole process should be fast. Remember, the code change doesn’t have to be perfect!

> 📢 If we are too busy to do a full review, we can still send a quick response that lets the developer know when we will get to it. Or we can suggest another reviewer! We should delegate, and don’t allow the code change request to sour for days.

Reviewer: ```suggestion
:::note
If we are too busy to do a full review, we can still send a quick response that lets the developer know when we will get to it. Or we can suggest another reviewer! We should delegate, and don’t allow the code change request to sour for days.
:::
```

#### Large code change requests

Are unacceptable and reviewers have the right to break the fast code review rule. Usually the way-to-go is to ask the author to split the code change into smaller requests (all branches should be branched from single/main branch for that feature). This way, all requests can be reviewed in a single day and when they're merged, the feature is complete and large code change is more-or-less already reviewed, so we only need to do a simple final check.

### How to write code review comments

#### 3 magic questions we should always ask ourselves

Always:

1. is it kind
2. is it an absolute truth
3. is it necessary

#### Courtesy

> In general, it is important to be [courteous and respectful](https://chromium.googlesource.com/chromium/src/+/master/docs/cr_respect.md) while also being very clear and helpful to the developer whose code we are reviewing. One way to do this is to be sure that we are always making comments about the code and never making comments about the developer.

Reviewer: maybe drop the link as it relates to Google culture

Author: but is it that tightly coupled with it? because it's just 8 DOs and 4 DON'Ts which seem pretty generic?

Reviewer 2: In any case it should.prpbably be permalink 🙂

Author: agreed, but i think this is the only way it's usable (they use it too) :sadge:

Author: i mean, i guess i can extract those dos and donts to this doc if we don't want to depend on Google haha

Explanations written only in the code review tool are not helpful to future code readers. Except in situations where we are unfamiliar with the code, and others are.

### Handling pushback in code reviews

#### Who is right

When a developer disagrees with our suggestions, we should take a moment to consider if they are correct. Often, they are closer to the code than we are, so they might really have a better insight. If we still think that we are right, we should explain further why. We should just remember to be polite and mature.

#### Cleaning it up later

Often, authors will ask if they can clean something later. In the following days, or next code change requests. This really depends on what your process looks like and how strict it is. If it is not strict, this will rarely happen. We should be cautious when approaching will-do-it-later actions.

### Resolving conversations

The reviewer who started the conversation must be the one who resolves it. **Nobody else** can resolve the conversation.

### References

https://google.github.io/eng-practices/review/reviewer/

https://phauer.com/2018/code-review-guidelines/#be-humble

> ## Author guide

Reviewer: Redunant, would remove

### How to handle reviewer comments

#### We should not take it personally

The goal of a review is to maintain the quality of our codebase and our products. When a reviewer provides a critique of your code, we should think of it as their attempt to help us, the codebase, and ChainSafe, rather than as a personal attack on us or our abilities.

📢 We should **never respond in anger to code review comments.** That is a serious breach of professional etiquette that will live forever in the code review tool. Even if reviewers expressed their frustration first.

#### We should fix the code before commenting

If a reviewer says that they don’t understand something in our code, our first response should be to clarify the code itself! If the code can’t be clarified, we should add a code comment that explains why the code is there. If a comment seems pointless, only then should our response be an explanation in the code review tool.

#### Authors can never ever resolve conversations

> Only the original reviewer can resolve the conversation.

Reviewer: ```suggestion
Only the original reviewer who wrote the comment can resolve the conversation.
```

## Changelog

> Assuming your project has releases, and is continually making updates, the project will need a changelog. When submitting a pr make sure to update the changelog reflecting the changes that your PR is making. This way, when a new release is made, there is already a changelog ready to go and no extra work is needed.

Reviewer 2: we should modify this, on most projects it's autogenerated

## Reviewer guide

The guide for reviewers is mostly a condensed version of [Google’s code review best practices](https://google.github.io/eng-practices/review/reviewer/). Some parts are just c/p, some parts are emphasized, some things are left out, and some are added or are different.

### The goal of the code review

> In general, reviewers should favor approving a PR once it is in a state where it definitely improves the overall code health of the system being worked on, even if the PR isn’t perfect.

Reviewer 2: I would replace PR with "Code Change" everywhere

Author: https://github.com/ChainSafe/engineering-handbook/pull/107/commits/e85fb9323fcabd3ce964c1be910dd4aa9ed3d9f7
